src/cc_gpu_sketch_alg.cpp

```diff
@@ -0,0 +1,146 @@
+#include "cc_gpu_sketch_alg.h"
+
+#include <iostream>
+#include <vector>
+
+// Configure CCGPUSketchAlg
+void CCGPUSketchAlg::configure(CudaUpdateParams** _cudaUpdateParams, long* _sketchSeeds, int _num_host_threads) {
+  cudaUpdateParams = _cudaUpdateParams;
+  //sketches = _sketches;
+  sketchSeeds = _sketchSeeds;
+
+  num_buckets = cudaUpdateParams[0]->num_buckets;
+
+  num_device_threads = 1024;
+  num_device_blocks = 1;
+  num_host_threads = _num_host_threads;
+  batch_size = cudaUpdateParams[0]->batch_size;
+  stream_multiplier = cudaUpdateParams[0]->stream_multiplier;
+
+  // num_host_threads must be even number
+  if (num_host_threads % 2 != 0) {
+    std::cout << "num_host_threads must be even number!\n";
+    exit(EXIT_FAILURE);
+  }
+
+  // Initialize CUDA Streams
+  for (int i = 0; i < num_host_threads * stream_multiplier; i++) {
+    cudaStream_t stream;
+
+    cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking);
+
+    streams.push_back(stream);
+    streams_deltaApplied.push_back(1);
+    streams_src.push_back(-1);
+    streams_num_graphs.push_back(-1);
+  }
+  
+  // Finished configuration, turn on the flag
+  isConfigured = true;
+}
+
+void CCGPUSketchAlg::apply_update_batch(int thr_id, node_id_t src_vertex,
+                                     const std::vector<node_id_t> &dst_vertices) {
+  if (update_locked) throw UpdateLockedException();
+  if (!isConfigured) {
+    std::cout << "CCGPUSketchAlg has not been configured!\n";
+  }
+  /*Sketch &delta_sketch = *delta_sketches[thr_id];
+  delta_sketch.zero_contents();
+
+  for (const auto &dst : dst_vertices) {
+    delta_sketch.update(static_cast<vec_t>(concat_pairing_fn(src_vertex, dst)));
+  }
+
+  std::unique_lock<std::mutex>(sketches[src_vertex]->mutex);
+  sketches[src_vertex]->merge(delta_sketch);*/
+
+  int stream_id = thr_id * stream_multiplier;
+  int stream_offset = 0;
+  while(true) {
+    if (cudaStreamQuery(streams[stream_id + stream_offset]) == cudaSuccess) {
+      // Update stream_id
+      stream_id += stream_offset;
+
+      // CUDA Stream is available, check if it has any delta sketch
+      if(streams_deltaApplied[stream_id] == 0) {
+        streams_deltaApplied[stream_id] = 1;
+
+        // Transfer delta sketch from GPU to CPU
+        cudaMemcpyAsync(&cudaUpdateParams[0]->h_bucket_a[stream_id * num_buckets], &cudaUpdateParams[0]->d_bucket_a[stream_id * num_buckets], num_buckets * sizeof(vec_t), cudaMemcpyDeviceToHost, streams[stream_id]);
```

Comment:
Hmm, based on my experience so far, I try to avoid transferring objects (in this case, it will be `Buckets`) between CPU and GPU. That's because you can't simply copy objects between CPU-GPU, you have to perform a deep copy of objects, meaning you have to manually call the `cudaMemcpy()` the contents of the objects. This means I have to call `cudaMemcpy()` all the buckets' alpha and gamma which is basically the same as having two separate arrays of `bucket_a` and `bucket_c` 

Here is a good example of what is required to copy an object between CPU and GPU: https://stackoverflow.com/questions/49947149/cuda-object-copy-from-device-to-host

Reply:
I see. This is a good reason to be cautious. However, in this case I believe this should be fine as the Bucket struct does not contain any pointers or references. In other words the Bucket is a 1 dimensional object.

In pure CPU terms I could perform a simple `memcpy()` between two Bucket arrays and expect the Bucket contents to be transferred correct. I believe this would generalize to CPU->GPU or GPU->CPU `cudaMemcpy`s.

I suppose it likely doesn't make that much of a difference in code elegance or performance. So either way is probably fine.